internal/terraform/node_resource_plan.go

```diff
@@ -116,9 +114,124 @@ func (n *nodeExpandPlannableResource) DynamicExpand(ctx EvalContext) (*Graph, tf
+		return g, diags
+	}
+
+	g, expandDiags := n.dynamicExpand(ctx, moduleInstances, imports)
+	diags = diags.Append(expandDiags)
+	return g, diags
+}
+
+// Import blocks are expanded in conjunction with their associated resource block.
+func (n *nodeExpandPlannableResource) expandResourceImports(ctx EvalContext) (addrs.Map[addrs.AbsResourceInstance, string], tfdiags.Diagnostics) {
+	// Imports maps the target address to an import ID.
+	imports := addrs.MakeMap[addrs.AbsResourceInstance, string]()
+	var diags tfdiags.Diagnostics
+
+	if len(n.importTargets) == 0 {
+		return imports, diags
+	}
+
+	// Import blocks are only valid within the root module, and must be
+	// evaluated within that context
+	ctx = evalContextForModuleInstance(ctx, addrs.RootModuleInstance)
+
+	for _, imp := range n.importTargets {
+		if imp.Config == nil {
+			// if we have a legacy addr, it was supplied on the commandline so
+			// there is nothing to expand
+			if !imp.LegacyAddr.Equal(addrs.AbsResourceInstance{}) {
+				imports.Put(imp.LegacyAddr, imp.IDString)
+				return imports, diags
+			}
+
+			// legacy import tests may have no configuration
+			log.Printf("[WARN] no configuration for import target %#v", imp)
+			continue
+		}
+
+		if imp.Config.ForEach == nil {
+			importID, evalDiags := evaluateImportIdExpression(imp.Config.ID, ctx, EvalDataForNoInstanceKey)
+			diags = diags.Append(evalDiags)
+			if diags.HasErrors() {
+				return imports, diags
+			}
+
+			traversal, hds := hcl.AbsTraversalForExpr(imp.Config.To)
+			diags = diags.Append(hds)
+			to, tds := addrs.ParseAbsResourceInstance(traversal)
+			diags = diags.Append(tds)
+			if diags.HasErrors() {
+				return imports, diags
+			}
+
+			imports.Put(to, importID)
+
+			log.Printf("[TRACE] expandResourceImports: found single import target %s", to)
+			continue
+		}
+
+		forEachData, forEachDiags := newForEachEvaluator(imp.Config.ForEach, ctx, false).ImportValues()
+		diags = diags.Append(forEachDiags)
+		if forEachDiags.HasErrors() {
+			return imports, diags
+		}
+
+		for _, keyData := range forEachData {
+			res, evalDiags := evalImportToExpression(imp.Config.To, keyData)
+			diags = diags.Append(evalDiags)
+			if diags.HasErrors() {
+				return imports, diags
+			}
+
+			importID, evalDiags := evaluateImportIdExpression(imp.Config.ID, ctx, keyData)
+			diags = diags.Append(evalDiags)
+			if diags.HasErrors() {
+				return imports, diags
+			}
+
+			imports.Put(res, importID)
+			log.Printf("[TRACE] expandResourceImports: expanded import target %s", res)
+		}
+	}
+
+	// filter out any import which already exist in state
+	state := ctx.State()
+	for _, el := range imports.Elements() {
+		if state.ResourceInstance(el.Key) != nil {
+			log.Printf("[DEBUG] expandResourceImports: skipping import address %s already in state", el.Key)
+			imports.Remove(el.Key)
+		}
+	}
+
+	return imports, diags
+}
+
+// validateExpandedImportTargets checks that all expanded imports correspond to
+// a configured instance.
```

Comment:
When I was tracing the parallel code paths, I expected this to be called by both paths... but really this is just the non-deferring version, and the deferrable version of the logic is inlined around line 107 in node_resource_partial_plan. Might be worth noting that for the next reader, given how tricky I'm finding it to trace the parallel logic. 

```suggestion
// validateExpandedImportTargets checks that all expanded imports correspond to
// a configured instance. This is only used in dynamicExpand; dynamicExpandPartial 
// has its own inlined version of this logic.
```

Reply:
Done, I've resolved this as suggested! I didn't take your suggestion directly just cos I made other changes in response to other comments at the same time and didn't want to deal with merging and stuff.